File: clickstream-analytics/cdk/lib/clickstream-stack.ts
import { Stack, StackProps, Duration, RemovalPolicy, CfnOutput } from "aws-cdk-lib";
import { Construct } from "constructs";
import * as kms from "aws-cdk-lib/aws-kms";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as iam from "aws-cdk-lib/aws-iam";
import * as logs from "aws-cdk-lib/aws-logs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as apigw from "aws-cdk-lib/aws-apigateway";
import * as wafv2 from "aws-cdk-lib/aws-wafv2";
import * as kds from "aws-cdk-lib/aws-kinesis";
import * as firehose from "aws-cdk-lib/aws-kinesisfirehose";
import * as s3dest from "aws-cdk-lib/aws-kinesisfirehose-destinations";
import * as eventsrc from "aws-cdk-lib/aws-lambda-event-sources";
import { ClickstreamVpc } from "./constructs/vpc";
import { ClickstreamIam } from "./constructs/iam";
import { ClickstreamGlue } from "./constructs/glue";
import { ClickstreamAthena } from "./constructs/athena";
import { ClickstreamOpenSearch } from "./constructs/opensearch";

export class ClickstreamStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const vpc = new ClickstreamVpc(this, "Net").vpc;
    const key = new kms.Key(this, "DataKey", { enableKeyRotation: true, alias: "alias/clickstream-data" });

    const processed = new s3.Bucket(this, "Processed", {
      encryption: s3.BucketEncryption.KMS, encryptionKey: key, blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL, enforceSSL: true,
      lifecycleRules: [{ expiration: Duration.days(365) }], removalPolicy: RemovalPolicy.DESTROY, autoDeleteObjects: true
    });
    const bad = new s3.Bucket(this, "Bad", {
      encryption: s3.BucketEncryption.KMS, encryptionKey: key, blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL, enforceSSL: true,
      lifecycleRules: [{ expiration: Duration.days(30) }], removalPolicy: RemovalPolicy.DESTROY, autoDeleteObjects: true
    });

    const stream = new kds.Stream(this, "EventStream", {
      shardCount: 1, streamMode: kds.StreamMode.PROVISIONED, encryption: kds.StreamEncryption.KMS, encryptionKey: key,
      retentionPeriod: Duration.hours(24)
    });

    const ingest = new lambda.Function(this, "IngestFn", {
      runtime: lambda.Runtime.NODEJS_20_X, handler: "index.handler", code: lambda.Code.fromAsset("../lambda/ingest"),
      memorySize: 256, timeout: Duration.seconds(10),
      environment: { STREAM_NAME: stream.streamName, BAD_BUCKET: bad.bucketName, ALLOW_ORIGINS: "https://localhost,https://example.com" },
      logRetention: logs.RetentionDays.ONE_WEEK
    });
    stream.grantWrite(ingest);
    bad.grantWrite(ingest);

    const api = new apigw.RestApi(this, "Api", { deployOptions: { stageName: "prod", loggingLevel: apigw.MethodLoggingLevel.INFO } });
    api.root.addResource("events").addMethod("POST", new apigw.LambdaIntegration(ingest));

    const webAcl = new wafv2.CfnWebACL(this, "ApiWaf", {
      defaultAction: { allow: {} }, scope: "REGIONAL", name: "clickstream-api-waf",
      visibilityConfig: { cloudWatchMetricsEnabled: true, metricName: "api-waf", sampledRequestsEnabled: true },
      rules: [{ name: "ManagedCommon", priority: 1, statement: { managedRuleGroupStatement: { vendorName: "AWS", name: "AWSManagedRulesCommonRuleSet" } },
        visibilityConfig: { cloudWatchMetricsEnabled: true, metricName: "common", sampledRequestsEnabled: true }, overrideAction: { none: {} } }]
    });
    new wafv2.CfnWebACLAssociation(this, "WafAssoc", { resourceArn: api.deploymentStage.stageArn, webAclArn: webAcl.attrArn });

    const fhRole = new iam.Role(this, "FirehoseRole", { assumedBy: new iam.ServicePrincipal("firehose.amazonaws.com") });
    const delivery = new firehose.DeliveryStream(this, "Delivery", {
      destinations: [ new s3dest.S3Bucket(processed, { bufferingInterval: Duration.seconds(60), dataOutputPrefix: "year=!{timestamp:yyyy}/month=!{timestamp:MM}/day=!{timestamp:dd}/hour=!{timestamp:HH}/",
        errorOutputPrefix: "errors/!{firehose:error-output-type}/", compression: s3dest.Compression.GZIP, logging: true, role: fhRole }) ],
      encryption: firehose.StreamEncryption.KMS, encryptionKey: key
    });
    new ClickstreamIam(this, "Iam", processed, key, fhRole);

    // Stream processor Lambda in VPC to forward to Firehose (and optionally OS)
    const sp = new lambda.Function(this, "StreamProcessor", {
      runtime: lambda.Runtime.NODEJS_20_X, handler: "index.handler", code: lambda.Code.fromAsset("../lambda/stream-processor"),
      memorySize: 512, timeout: Duration.seconds(30), vpc, vpcSubnets: { subnets: vpc.privateSubnets },
      environment: { FIREHOSE_NAME: delivery.deliveryStreamName }
    });
    stream.grantRead(sp);
    delivery.grantPutRecords(sp);
    sp.addEventSource(new eventsrc.KinesisEventSource(stream, { batchSize: 100, startingPosition: lambda.StartingPosition.LATEST }));

    // Glue + Athena
    new ClickstreamGlue(this, "Glue", processed);
    new ClickstreamAthena(this, "Athena", processed);

    // OpenSearch (optional; VPC only)
    new ClickstreamOpenSearch(this, "OS", vpc);

    new CfnOutput(this, "ApiUrl", { value: api.urlForPath("/events") });
    new CfnOutput(this, "ProcessedBucket", { value: processed.bucketName });
  }
}

File: clickstream-analytics/cdk/clickstream_stack.py
from aws_cdk import (
    Stack, Duration, RemovalPolicy, CfnOutput,
    aws_apigateway as apigw,
    aws_cloudwatch as cw,
    aws_ec2 as ec2,
    aws_firehose as firehose,
    aws_glue as glue,
    aws_iam as iam,
    aws_kinesis as kinesis,
    aws_kms as kms,
    aws_lambda as _lambda,
    aws_lambda_event_sources as event_sources,
    aws_logs as logs,
    aws_opensearchservice as opensearch,
    aws_s3 as s3,
    aws_s3_deployment as s3deploy,
    aws_sqs as sqs,
    aws_wafv2 as wafv2,
)
from constructs import Construct
from pathlib import Path

class ClickstreamStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, **kwargs):
        super().__init__(scope, construct_id, **kwargs)

        key = kms.Key(self, "DataKey", enable_key_rotation=True)

        vpc = ec2.Vpc(
            self, "AnalyticsVpc",
            max_azs=2,
            nat_gateways=1,
            subnet_configuration=[
                ec2.SubnetConfiguration(name="public", subnet_type=ec2.SubnetType.PUBLIC, cidr_mask=24),
                ec2.SubnetConfiguration(name="private", subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS, cidr_mask=24),
            ],
        )

        stream = kinesis.Stream(
            self, "EventStream",
            stream_mode=kinesis.StreamMode.ON_DEMAND,
            encryption=kinesis.StreamEncryption.KMS,
            encryption_key=key,
            retention_period=Duration.hours(48),
        )

        lake = s3.Bucket(
            self, "ProcessedLake",
            encryption=s3.BucketEncryption.KMS,
            encryption_key=key,
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
            versioned=True,
            enforce_ssl=True,
            lifecycle_rules=[
                s3.LifecycleRule(
                    id="curated-retention",
                    enabled=True,
                    transitions=[s3.Transition(storage_class=s3.StorageClass.INFREQUENT_ACCESS, transition_after=Duration.days(30))],
                    expiration=Duration.days(395),
                )
            ],
            removal_policy=RemovalPolicy.RETAIN,
        )

        error_bucket = s3.Bucket(
            self, "ErrorBucket",
            encryption=s3.BucketEncryption.KMS,
            encryption_key=key,
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
            enforce_ssl=True,
            lifecycle_rules=[s3.LifecycleRule(expiration=Duration.days(30))],
            removal_policy=RemovalPolicy.RETAIN,
        )

        dlq = sqs.Queue(
            self, "ProcessorDlq",
            encryption=sqs.QueueEncryption.KMS,
            encryption_master_key=key,
            retention_period=Duration.days(14),
        )

        # Glue catalog/table for Parquet lake.
        database = glue.CfnDatabase(
            self, "GlueDatabase",
            catalog_id=self.account,
            database_input=glue.CfnDatabase.DatabaseInputProperty(name="clickstream"),
        )

        table = glue.CfnTable(
            self, "GlueTable",
            catalog_id=self.account,
            database_name="clickstream",
            table_input=glue.CfnTable.TableInputProperty(
                name="events",
                table_type="EXTERNAL_TABLE",
                parameters={"classification":"parquet"},
                partition_keys=[
                    glue.CfnTable.ColumnProperty(name="event_date", type="string"),
                    glue.CfnTable.ColumnProperty(name="event_hour", type="string"),
                ],
                storage_descriptor=glue.CfnTable.StorageDescriptorProperty(
                    location=f"s3://{lake.bucket_name}/events/",
                    input_format="org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat",
                    output_format="org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat",
                    serde_info=glue.CfnTable.SerdeInfoProperty(
                        serialization_library="org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"
                    ),
                    columns=[
                        glue.CfnTable.ColumnProperty(name="event_id",type="string"),
                        glue.CfnTable.ColumnProperty(name="event_time",type="timestamp"),
                        glue.CfnTable.ColumnProperty(name="session_id",type="string"),
                        glue.CfnTable.ColumnProperty(name="anonymous_id",type="string"),
                        glue.CfnTable.ColumnProperty(name="user_id",type="string"),
                        glue.CfnTable.ColumnProperty(name="event_type",type="string"),
                        glue.CfnTable.ColumnProperty(name="page_url",type="string"),
                        glue.CfnTable.ColumnProperty(name="device_type",type="string"),
                        glue.CfnTable.ColumnProperty(name="source",type="string"),
                        glue.CfnTable.ColumnProperty(name="revenue",type="double"),
                    ],
                ),
            ),
        )
        table.add_dependency(database)

        firehose_role = iam.Role(
            self, "FirehoseRole",
            assumed_by=iam.ServicePrincipal("firehose.amazonaws.com"),
        )
        lake.grant_read_write(firehose_role)
        error_bucket.grant_read_write(firehose_role)
        key.grant_encrypt_decrypt(firehose_role)

        delivery = firehose.CfnDeliveryStream(
            self, "DeliveryStream",
            delivery_stream_type="DirectPut",
            extended_s3_destination_configuration=firehose.CfnDeliveryStream.ExtendedS3DestinationConfigurationProperty(
                bucket_arn=lake.bucket_arn,
                role_arn=firehose_role.role_arn,
                prefix="events/event_date=!{timestamp:yyyy-MM-dd}/event_hour=!{timestamp:HH}/",
                error_output_prefix="errors/!{firehose:error-output-type}/",
                buffering_hints=firehose.CfnDeliveryStream.BufferingHintsProperty(interval_in_seconds=300,size_in_m_bs=64),
                compression_format="GZIP",
                encryption_configuration=firehose.CfnDeliveryStream.EncryptionConfigurationProperty(
                    kms_encryption_config=firehose.CfnDeliveryStream.KMSEncryptionConfigProperty(awskms_key_arn=key.key_arn)
                ),
            ),
        )

        ingest = _lambda.Function(
            self, "IngestFunction",
            runtime=_lambda.Runtime.PYTHON_3_11,
            handler="handler.handler",
            code=_lambda.Code.from_asset("../lambda/ingest"),
            timeout=Duration.seconds(10),
            memory_size=256,
            environment={"STREAM_NAME": stream.stream_name},
            log_retention=logs.RetentionDays.ONE_MONTH,
        )
        stream.grant_write(ingest)

        domain = opensearch.Domain(
            self, "SearchDomain",
            version=opensearch.EngineVersion.OPENSEARCH_2_13,
            vpc=vpc,
            vpc_subnets=[ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS)],
            capacity=opensearch.CapacityConfig(data_nodes=2, data_node_instance_type="t3.small.search"),
            ebs=opensearch.EbsOptions(enabled=True, volume_size=50),
            node_to_node_encryption=True,
            encryption_at_rest=opensearch.EncryptionAtRestOptions(enabled=True,kms_key=key),
            enforce_https=True,
            zone_awareness=opensearch.ZoneAwarenessConfig(enabled=True,availability_zone_count=2),
            removal_policy=RemovalPolicy.RETAIN,
        )

        processor = _lambda.Function(
            self, "ProcessorFunction",
            runtime=_lambda.Runtime.PYTHON_3_11,
            handler="handler.handler",
            code=_lambda.Code.from_asset("../lambda/processor"),
            timeout=Duration.seconds(60),
            memory_size=1024,
            environment={
                "DELIVERY_STREAM": delivery.ref,
                "OS_HOST": domain.domain_endpoint,
                "OS_INDEX": "clickstream-events",
            },
            vpc=vpc,
            log_retention=logs.RetentionDays.ONE_MONTH,
            dead_letter_queue=dlq,
        )
        processor.add_event_source(
            event_sources.KinesisEventSource(
                stream,
                starting_position=_lambda.StartingPosition.LATEST,
                batch_size=200,
                bisect_batch_on_error=True,
                retry_attempts=5,
                parallelization_factor=2,
                report_batch_item_failures=True,
            )
        )
        processor.add_to_role_policy(iam.PolicyStatement(
            actions=["firehose:PutRecord","firehose:PutRecordBatch"],
            resources=[delivery.attr_arn],
        ))
        domain.grant_write(processor)
        key.grant_encrypt_decrypt(processor)

        api = apigw.RestApi(
            self, "EventApi",
            deploy_options=apigw.StageOptions(
                stage_name="prod",
                throttling_rate_limit=6000,
                throttling_burst_limit=10000,
                metrics_enabled=True,
                logging_level=apigw.MethodLoggingLevel.ERROR,
            ),
        )
        events_resource = api.root.add_resource("events")
        events_resource.add_method("POST", apigw.LambdaIntegration(ingest))

        web_acl = wafv2.CfnWebACL(
            self, "WebAcl",
            scope="REGIONAL",
            default_action=wafv2.CfnWebACL.DefaultActionProperty(allow={}),
            visibility_config=wafv2.CfnWebACL.VisibilityConfigProperty(
                cloud_watch_metrics_enabled=True,
                metric_name="clickstream-web-acl",
                sampled_requests_enabled=True,
            ),
            rules=[
                wafv2.CfnWebACL.RuleProperty(
                    name="AWSManagedCommon",
                    priority=0,
                    override_action=wafv2.CfnWebACL.OverrideActionProperty(none={}),
                    statement=wafv2.CfnWebACL.StatementProperty(
                        managed_rule_group_statement=wafv2.CfnWebACL.ManagedRuleGroupStatementProperty(
                            name="AWSManagedRulesCommonRuleSet",
                            vendor_name="AWS",
                        )
                    ),
                    visibility_config=wafv2.CfnWebACL.VisibilityConfigProperty(
                        cloud_watch_metrics_enabled=True,
                        metric_name="common-rules",
                        sampled_requests_enabled=True,
                    ),
                ),
                wafv2.CfnWebACL.RuleProperty(
                    name="RateLimit",
                    priority=1,
                    action=wafv2.CfnWebACL.RuleActionProperty(block={}),
                    statement=wafv2.CfnWebACL.StatementProperty(
                        rate_based_statement=wafv2.CfnWebACL.RateBasedStatementProperty(limit=12000,aggregate_key_type="IP")
                    ),
                    visibility_config=wafv2.CfnWebACL.VisibilityConfigProperty(
                        cloud_watch_metrics_enabled=True,
                        metric_name="rate-limit",
                        sampled_requests_enabled=True,
                    ),
                ),
            ],
        )

        wafv2.CfnWebACLAssociation(
            self, "ApiWafAssociation",
            resource_arn=f"arn:aws:apigateway:{self.region}::/restapis/{api.rest_api_id}/stages/prod",
            web_acl_arn=web_acl.attr_arn,
        )

        cw.Alarm(
            self, "IteratorAgeAlarm",
            metric=stream.metric_get_records_iterator_age_milliseconds(period=Duration.minutes(1)),
            threshold=60000,
            evaluation_periods=3,
        )
        cw.Alarm(
            self, "ProcessorErrorsAlarm",
            metric=processor.metric_errors(period=Duration.minutes(5)),
            threshold=5,
            evaluation_periods=2,
        )

        CfnOutput(self, "ApiUrl", value=api.url_for_path("/events"))
        CfnOutput(self, "KinesisStreamName", value=stream.stream_name)
        CfnOutput(self, "ProcessedBucketName", value=lake.bucket_name)
        CfnOutput(self, "OpenSearchEndpoint", value=domain.domain_endpoint)
